INT-3752: Add AggregatingMetricsFactory

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/INT-3752
JIRA: https://jira.spring.io/browse/INT-3754

The statistics are based on this count so, for example the
`meanSendDuration` reflects the mean of total duration from the first send to
the last in each batch.

Also change the linked lists to `Deque`s.

Also remove `isTraceEnabled()` calls - too expensive in high volume environments.

garyrussell committed:
INT-3752: Add AggregatingMetricsFactory

JIRA: https://jira.spring.io/browse/INT-3752
JIRA: https://jira.spring.io/browse/INT-3754

The statistics are based on this count so, for example the
`meanSendDuration` reflects the mean of total duration from the first send to
the last in each batch.

Also change the linked lists to `Deque`s.

Also remove `isTraceEnabled()` calls - too expensive in high volume environments.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/channel/management/AggregatingMessageChannelMetrics.java:
```diff
@@ -0,0 +1,98 @@
+	 * @param sendSuccessRatio an {@link ExponentialMovingAverageRatio} for calculating the success ratio.
+	 * @param sendRate an {@link ExponentialMovingAverageRate} for calculating the send rate.
+	 * @param sampleSize the sample size
+	 * @since 4.2
```
Minor: this is redundant here, since the whole class is new.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/channel/management/AggregatingMessageChannelMetrics.java:
```diff
@@ -0,0 +1,98 @@
+import org.springframework.integration.support.management.MetricsContext;
+
+/**
+ * Implementation that provides the total response times over a sample, to avoid fetching the system time
```
`Implementation` of what?
Maybe this one would be better?

> The {@link DefaultMessageChannelMetrics} extension ...

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/channel/management/AggregatingMessageChannelMetrics.java:
```diff
@@ -0,0 +1,98 @@
+	@Override
+	public synchronized MetricsContext beforeSend() {
+		long start = 0;
+		if (this.sendCount.getAndIncrement() % this.sampleSize == 0 && isFullStatsEnabled()) {
```
maybe reverse?
```
(isFullStatsEnabled() && this.sendCount.getAndIncrement() % this.sampleSize == 0)
```
Because we always `getAndIncrement()` independently of full stat enablement.
Have I missed anything?

garyrussell replied:
?? I'm taking advantage of java `if` short circuiting - we want `sendCount` to always be incremented. With your version, the count will only be incremented if `isFullStatsEnabled()`.

With my version, the count is always incremented and the `ifFullStatsEnabled()` is only invoked if the mod condition is true.

garyrussell replied:
However, I just tested this...

		long count = this.sendCount.getAndIncrement();
		if (isFullStatsEnabled() && count % this.sampleSize == 0) {
			start = System.nanoTime();
			this.sendRate.increment(start);
		}

and it's 35% faster - storing the temporary result is much faster than the `%` operation.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/management/AggregatingMessageHandlerMetrics.java:
```diff
@@ -0,0 +1,86 @@
+		}
+	}
+
+}
```
The same concerns to this class, too.

garyrussell replied:
And the same answer applies :smile:

artembilan commented:
That's all to this PR.
Just let me know and I can address all of them on merge.

garyrussell committed:
Polishing

Assign the count to a local variable so the mod (%) operation is short
circuited rather than short circuiting the call to isFullStatsEnabled().

garyrussell commented:
Pushed

garyrussell committed:
INT-3752: Fix Aggregating Metrics

Previously, the duration was for the first message in each sample.

We need to capture the total elapsed time for the sample and use it
for the duration calculation.

Add 'newCount' to the context so the after... method can calculate
the duration at the end of the sample.

The start field does not need to be volatile because its updates
are synchronized.

garyrussell commented:
However, you asking this question made me realize the calculations were completely wrong.

Fixed.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/management/DefaultMessageHandlerMetrics.java:
```diff
@@ -158,9 +154,9 @@ public Statistics getDuration() {
 	}
 
-	private static class DefaultHandlerMetricsContext implements MetricsContext {
+	public static class DefaultHandlerMetricsContext implements MetricsContext {
```
It's very aggressive: `protected` works well.

artembilan replied:
Fixing during merge...

garyrussell replied:
I pushed it with the javadocs

garyrussell committed:
INT-3752: Fix Javadocs

garyrussell commented:
Pushed

artembilan commented:
Merged after squash and rebase
